main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala

```diff
@@ -1555,26 +1555,51 @@ object Weeder2 {
-      def visitCase(caze: Tree): Validation[(Name.Ident, Expr), CompilationMessage] =
-        mapN(pickNameIdent(caze), pickExpr(caze)) {
-          case (id, e) => (id, e)
-        }
+    private def visitExtMatch(tree: Tree)(implicit sctx: SharedContext): Validation[Expr, CompilationMessage] = {
+      expect(tree, TreeKind.Expr.ExtMatch)
+      val rules0 = pickAll(TreeKind.Expr.ExtMatchRuleFragment, tree)
+      flatMapN(pickExpr(tree), traverse(rules0)(visitExtMatchRule)) {
+        // Case: no valid match rule found in ematch expr
+        case (expr, Nil) =>
+          val error = NeedAtleastOne(NamedTokenSet.ExtMatchRule, SyntacticContext.Expr.OtherExpr, loc = expr.loc)
+          // Fall back on Expr.Error. Parser has reported an error here.
+          Validation.Success(Expr.Error(error))
 
-      expect(tree, TreeKind.Expr.ExtensibleMatch)
+        case (expr, rules) if rules.length == 2 => // Check for exactly 2 to prevent crash when desugaring in Kinder
+          Validation.Success(Expr.ExtMatch(expr, rules, tree.loc))
 
-      mapN(pickExpr(tree), pickNameIdent(tree), traverse(pickAll(TreeKind.Case, tree))(visitCase)) {
-        case (exp1, label, (ident2, exp2) :: (ident3, exp3) :: Nil) =>
-          Expr.ExtensibleMatch(label, exp1, ident2, exp2, ident3, exp3, tree.loc)
-        case _ =>
-          throw InternalCompilerException("Illegal extensible match", tree.loc)
+        case (expr, rules) =>
+          val error = Malformed(NamedTokenSet.ExtMatchRule, SyntacticContext.Expr.OtherExpr, loc = expr.loc)
+          sctx.errors.add(error)
+          Validation.Success(Expr.Error(error))
       }
     }
 
-    private def visitExtensibleTag(tree: Tree)(implicit sctx: SharedContext): Validation[Expr, CompilationMessage] = {
-      expect(tree, TreeKind.Expr.ExtensibleTag)
+    private def visitExtMatchRule(tree: Tree)(implicit sctx: SharedContext): Validation[ExtMatchRule, CompilationMessage] = {
+      expect(tree, TreeKind.Expr.ExtMatchRuleFragment)
+      val exprs = pickAll(TreeKind.Expr.Expr, tree)
+      flatMapN(Patterns.pickExtPattern(tree), traverse(exprs)(visitExpr)) {
+        // case pattern => expr
+        case ((label, List(ExtPattern.Var(ident, loc))), expr :: Nil) =>
+          Validation.Success(ExtMatchRule(label, List(ExtPattern.Var(ident, loc)), expr, tree.loc))
+
+        // case pattern => expr
+        case ((label, List(ExtPattern.Wild(loc))), expr :: Nil) =>
+          Validation.Success(ExtMatchRule(label, List(ExtPattern.Wild(loc)), expr, tree.loc))
+
+        // Fall back on Expr.Error. Parser has reported an error here.
+        case ((label, _), _) =>
+          val error = Malformed(NamedTokenSet.ExtMatchRule, SyntacticContext.Expr.OtherExpr, loc = tree.loc)
+          sctx.errors.add(error)
+          Validation.Failure(error) // Hard failure to prevent crash when desugaring in Kinder
```

Comment:
What is this?

Reply:
This is just to be safe until we fully support all ext patterns and arbitrary arity. I want to avoid crashing in the unsafe desugaring in Kinder